src/api.v2/helpers/worker/workSubmit/getExtraDependencies.js

```diff
@@ -1,12 +1,48 @@
 const workerVersions = require('../workerVersions');
+const Experiment = require('../../../model/Experiment');
+const getLastModified = require('../../s3/getLastModified');
+const bucketNames = require('../../../../config/bucketNames');
+const getS3Object = require('../../s3/getObject');
 
-const getClusteringSettings = async (message) => {
-  const { input: { config: { clusteringSettings } } } = message;
+const getClusteringSettings = async (experimentId) => {
+  const {
+    processingConfig,
+  } = await new Experiment().findById(experimentId).first();
 
-  return clusteringSettings;
+  const {
+    configureEmbedding: { clusteringSettings: { method, methodSettings } },
+  } = processingConfig;
+
+  return { method, methodSettings: methodSettings[method] };
+};
+
+const getEmbeddingSettings = async (experimentId) => {
+  const {
+    processingConfig,
+  } = await new Experiment().findById(experimentId).first();
+
+  const {
+    configureEmbedding: {
+      embeddingSettings: { method, methodSettings },
+    },
+  } = processingConfig;
+
+  return { method, methodSettings: methodSettings[method] };
 };
 
-const getCellSetsThatAffectDownsampling = async (message, body, cellSets) => {
+const getCellSets = async (experimentId) => {
+  // TODO consider just fetching latest modified date instead of the whole object
+  const lastModified = await getLastModified({
+    Bucket: bucketNames.CELL_SETS,
+    Key: experimentId,
+  });
+
+  const lastVersion = `${bucketNames.CELL_SETS}/${experimentId}/${lastModified}`;
```

Comment:
if ``${bucketNames.CELL_SETS}/${experimentId}` is empty, because there are no cell sets, then the value of `lastModified` will be null. Then the value of `lastVersion` will be either something invalid, or line 40: `const lastVersion = `${bucketNames.CELL_SETS}/${experimentId}/${lastModified};` will generate an error. Right? Am I missing something?

Reply:
If there are no cellsets it will throw an error yes.  I think this is the desired behavior as you can't never send a work request unless the pipeline is done (and the cellsets are generated). I could do more defensive programming and return `lastModified=null` if the file does exist. It would make sense that, as long as it is still null, the dependency hasn't really been updated. I am not sure if I prefer that, or having the function fail fast because we assume there MUST be a cellset for work requests

Reply:
I also think that it is better to raise an error and fail fast, rather than try to hide it by making `lastModified=null` only to have a failure later in the code. With the current version in the code, theoretically there is way in which `lastModified` is `null` or `undefined` without any error being generated. This will happen if `metadata.LastModified` in this line: https://github.com/hms-dbmi-cellenics/api/pull/494/files#diff-a48ec42db65954d588678dca5b1e781ee3e0c643ed3ebe9abadd0c45ec64998fR12 Is it possible for `metadata.LastModified` to return `null` or `undefined`?

Reply:
I don't think so as it is a s3 attribute. If the file is not there it will fail